docs/parameterized-function-expressions.md

```diff
@@ -0,0 +1,179 @@
+local moneyBinding = React.createBinding(nil :: number?)
+```
+
+However, `::` has the ability to cast things in unsound manners. Both these calls are accepted with `::` when we would not want it in this case:
+
+```lua
+local function f(x: number?)
+    React.createBinding(x :: number) -- Allowed
+end
+
+type Object = { a: number }
+React.createBinding({} :: Object) -- Allowed
+```
+
+So while it would work in some cases, it provides an extra bit of uncertainty to the correctness of the call.
+
+In other cases, type parameters are used in places where there is no immediate value binding to that type. For example, a Signal type might be defined as:
+
+```lua
+type Signal<Args...> = {
+    fire: (Args...) -> (),
+    connect: ((Args...) -> ()) -> (),
+}
+
+local function createSignal<Args...>(): Signal<Args...>
+```
+
+The usual local assignment bindings work fine here:
+
+```lua
+local playerTookDamageSignal: Signal<Player, number> = createSignal()
+```
+
+...but only if this binding is actually declared through a local. In a case like:
+
+```lua
+Signals.playerTookDamageSignal = createSignal()
+```
+
+...we either have to put this out into another variable above, which increases surface area, or use `::`, which has the problems described previously. This is not a problem specific to function calls, but is one of the places where they would be useful.
+
+## Design
+
+We propose allowing the following:
+
+```lua
+-- Parameterized function call as statement
+f<<T, U>>()
+
+-- Parameterized function call as expression
+local a = f<<T, U>>()
+
+-- Calling through index
+local b = t.f<<T>>()
+local c = t:f<<T>>()
+local d = t["f"]<<T>>()
+
+-- Saving `f<<T>>` as a variable to called later
+local e = f<<T>>
+```
+
+For parsing, we will do this by extending `var` to allow `prefixexp '<<' TypeList '>>'`. This will bind the prefixexp equivalent to the substitution of the type parameters. That is to say...
+
+```lua
+local function f<T, U>() end
+
+-- Binds `T` and `U` to `number` and `string`
+local f1 = f<<number, string>>
+
+-- Binds just `T` to number. The unspecified `U` type is anonymous in the same way it would be if this were just `f`.
+local f2 = f<<number>>
+```
+
+When this new syntax is used on a value that is not a function, or does not take those types, a static-time error will be given. This syntax has zero runtime effect.
+
+## Drawbacks
+
+The syntax proposed in this RFC would close the door on the ability to add a bitwise left-shift operator to Luau. This is because this would make the following usage ambiguous:
+
+```lua
+-- Assuming we have right-shift, because there's no way we add one without the other
+return f<<a, b>>(c)
+```
+
+This is either a parameterized function call of `f(c)` with `a` and `b` as types, or a return of two values--`f << a`, and `b >> (c)`.
+
+Because there would be no ability to add bitwise shifts to the language, it is unlikely that any bitwise operators would be added, as it would be a stark omission.

```

Comment:
Should we see a RFC that basically puts the final nail in the coffin for bitwise ops? Time and time again, people want to see it, but `bit32` is better because it guarantees better performance characteristics anyway.